### DISCUSSION TITLE
[iOS] Shiny.Net.Http upload straight to onError

### DISCUSSION THREAD

**Discussion by GuidoNeele** [shinyorg/shiny#778] [Q&A] [answered]
I'm setting up background uploads using Shiny.Net.Http. Everything is working fine on Android but when uploading files on iOS the `IHttpTransferDelegate` `OnError` method is called straightaway with a "cancelled" exception (only a message no stacktrace). 

I've added the following code to `AppDelegate.cs`:
```csharp 
public override void HandleEventsForBackgroundUrl(UIApplication application, string sessionIdentifier, Action completionHandler)
{
      this.ShinyHandleEventsForBackgroundUrl(sessionIdentifier, completionHandler);
}
```

I think that's the only setup that needs to be done for iOS is that correct or am I missing something? 

Thanks in advance!

Guido

**Comment by aritchie** [comment]
Are you using the "Shiny" package to generate the AppDelegate?

**Reply by GuidoNeele** [reply]
No I'm not generating the AppDelegate. Haven't made the switch yet. So pasted in the lines of code specified on the documentation page (https://shinylib.net/httptransfers/getting-started). Will try the sample application today to see if that works for me.

**Reply by GuidoNeele** [reply]
Turns out PostData can't be null (https://stackoverflow.com/questions/38406303/urlsessionuploadtask-getting-automatically-cancelled-instantly/38512824#38512824). Made the following edit and now the server get's a request.

```csharp
HttpTransferRequest request = new HttpTransferRequest($"{host}/attachment", file.FullPath, true)
{
    HttpMethod = HttpMethod.Post,
    PostData = "{ 'empty': true }"
};
```

**Reply by aritchie** [reply]
The samples aren't done so don't use them as a standard just yet.  They are only in a state where they've been separated from the kitchen sink big sample.  The upload postdata being required is some sort of new requirement.  I can fix this internally

**Reply by GuidoNeele** [reply]
I've been doing some more testing and viewing source code and I don't think the file is being uploaded at the moment. Posted data to requestbin.com (pipedream) so I could see the posted data but the actual file is not being sent.

```
X-Amzn-Trace-Id: Root=1-6183c1d8-5ff95f5523134e703bbbca5f
Content-Length: 17
accept: */*
content-type: application/x-www-form-urlencoded
attachmentid: 082e7ee7-3142-4625-93de-84c26d94db6f
user-agent: REMOVED/3.6.27 CFNetwork/1312 Darwin/20.6.0
accept-language: en-US,en;q=0.9
accept-encoding: gzip, deflate, br

{ 'empty': true }
```

I'm not sure but I think the file should be added to the CreateUploadTask call.
https://github.com/shinyorg/shiny/blob/935eaecb0f4e8a6583d9e151cabeb9a05052d826/src/Shiny.Net.Http/Platforms/iOS/HttpTransferManager.cs#L65

Something like this:
```csharp
var fileUrl = new NSUrl(request.LocalFile.FullName);
var task = this.Session.CreateUploadTask(request.ToNative(), fileUrl);
var taskId = TaskIdentifier.Create(request.LocalFile);
task.TaskDescription = taskId.ToString();
var transfer = task.FromNative();
task.Resume();

return Task.FromResult(transfer);
```

I also noticed that `IHttpTransferDelegate.OnComplete` isn't called when the upload has succeeded. 
`OnComplete` is only called from `NSUrlSessionDownloadDelegate.DidFinishDownloading` and it looks like that's only used for downloads. Probably a check should be added to `NSUrlSessionDownloadDelegate.DidSendBodyData` or `NSUrlSessionDownloadDelegate.DidWriteData` to check if the task completed. Because I do get an event via `IHttpTransferManager.WhenUpdated()` with a `transfer` indicating status `Completed`.
https://github.com/shinyorg/shiny/blob/935eaecb0f4e8a6583d9e151cabeb9a05052d826/src/Shiny.Net.Http/Platforms/iOS/ShinyUrlSessionDelegate.cs#L82

**Comment by aritchie** [comment, marked as answer]
2.4 should resolve all of these issues as per #782 - thanks for all the help @GuidoNeele with this one!
